### PR TITLE
fix #2328 Handle safety flow through instanceof pattern matching

### LINKED ISSUE #2328: Extremely slow compilation for pattern matching instanceof method
## What happened?

I wrote code which looks roughly like:

```java
        @SuppressWarnings("CyclomaticComplexity")
        private static long getRoutingId(BlockId block) {
            if (block instanceof RDDBlockId b) {
                return b.rddId();
            } else if (block instanceof ShuffleBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleBlockBatchId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleBlockChunkId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleDataBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleIndexBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleChecksumBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShufflePushBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleMergedBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleMergedDataBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleMergedIndexBlockId b) {
                return b.shuffleId();
            } else if (block instanceof ShuffleMergedMetaBlockId b) {
                return b.shuffleId();
            } else if (block instanceof BroadcastBlockId b) {
                return b.broadcastId();
            } else {
                throw new SafeIllegalStateException("unhandled");
            }
```

where the underlying type of BlockId is a Scala case class with many subtypes. Not my favourite code, but surely close to as good as can be done.

Compilation is currently 9 minutes in. Stack traces are implying that the expensive code relates to the Safety Propagation code. JFR can be provided on request.

## What did you want to happen?

Compilation in a reasonable amount of time.

### PR DESCRIPTION
**Pull request by carterkozak**
==COMMIT_MSG==
Handle safety flow through instanceof pattern matching
==COMMIT_MSG==

### CODE REVIEW

File: baseline-error-prone/build.gradle [hunk 1]
```diff
@@ -3,6 +3,11 @@ import net.ltgt.gradle.errorprone.CheckSeverity
 apply plugin: 'java-library'
 apply plugin: 'com.palantir.external-publish-jar'
 
+javaVersions {
+    libraryTarget = 11
+    runtime = 17
+}
```
Comment (carterkozak) [inline baseline-error-prone/build.gradle]: We'll see what this breaks

File: baseline-refaster-rules/build.gradle [hunk 2]
```diff
@@ -34,4 +34,5 @@ moduleJvmArgs {
             'jdk.compiler/com.sun.tools.javac.tree',
             'jdk.compiler/com.sun.tools.javac.util'
     ]
+    opens = ['jdk.compiler/com.sun.tools.javac.comp']
```
Comment (fawind) [inline baseline-refaster-rules/build.gradle:37]: For my understanding, why do we need the additional "opens" declaration now? I don't see any new usage in the diff here.

**Review by fawind — APPROVED** [review]
👍

File: baseline-refaster-rules/build.gradle (hunk 2 above)
Comment (carterkozak) [reply baseline-refaster-rules/build.gradle:37]: This allows this module to pass tests using jdk17. Not required for the build, but initially I tried using jdk17 everywhere, but some tests are more difficult to fix.

**Pull request merged**